### DISCUSSION TITLE
How to access appsettings value in a typescript class?

### DISCUSSION THREAD

**Discussion by petebrieso** [serenity-is/Serenity#6468] [Help]
This isn't strictly a Serenity question, but I wonder if there's a cleaner way to approach this?
I've been using hacky method of putting values into global space in `_Layout.cshtml` like this
```
@using Microsoft.Extensions.Configuration
@inject IConfiguration Configuration
...
<script type="text/javascript">
  ...
    window["UploadBaseUrl"] = "@Configuration["UploadSettings:Url"]";
  
</script>
```
and then referencing that value with ` window["UploadBaseUrl"]`

surely a better method exists?

**Comment by stixoffire** [comment]
in pure c#  a static method could be used , but getting that in typecript you would need to add export class and register it in serenity. Caveat it is available to all ..

**Reply by petebrieso** [reply]
good idea, thanks. Do you happen to know the syntax to export /register to a ts unit?

**Reply by stixoffire** [reply]
If you have your typescript class I think just above it you use Serenity.Decorators register class. 
I did this along time ago (10 years maybe) with plain c# and as long as you can access the static method - you get the data again the caveat is it does not care about clients (who what they are all one to it) - the class is static . 
The static class as such is available regardless to the client .. all clients are accessing the same chunk of code. (no uniqueness) .. so you would need to determine the client if you want to change up for specific clients. An example for you - put a variable in the screen on one browser and view it form two different machines .. the result will come back with the value the static class has .. even if you allowed an update to a backing field both get the value.

I did a quick search 
https://hassantariqblog.wordpress.com/2015/09/12/appsettings-values-usage-through-static-class-c-2/

```
public class AppConfig
{
private static string _moduleTitle;
private static bool? _bundlingActive;
 
public static string ModuleTitle
{
    get
    {
        if (string.IsNullOrEmpty(_moduleTitle))
            _moduleTitle = ConfigurationManager.AppSettings["ModuleTitle"];
        return _moduleTitle;
    }
}
 
public static bool BundlingActive
{
    get
    {
        if (!_bundlingActive.HasValue)
            _bundlingActive = ConfigurationManager.AppSettings["BundlingActive"] == "1";
        return _bundlingActive.Value;
    }
}
}
```

**Reply by stixoffire** [reply]
Also look here

https://stackoverflow.com/questions/51926625/using-static-variables-in-c-sharp-web-application-accessing-session-variables



HttpContext is accessible only from things that are request specific, since it's a context of one and only request. And new controller instances are created by the framework for each request, with injected HttpContext. It's the developers job to pass it further if the need arises.

I recommend reading this article about it: https://dotnetcoretutorials.com/2017/01/05/accessing-httpcontext-asp-net-core/

    First in your startup.cs, you need to register IHttpContextAccessor as a service like so :
```

public void ConfigureServices(IServiceCollection services)
{
  services.AddMvc();
  services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
}
```

    When you create a helper/service class, you can then inject in the IHttpContextAccessor and use it. It would look like something not too dissimilar to this :

```
public class UserService : IUserService
{
  private readonly IHttpContextAccessor _httpContextAccessor;

  public UserService(IHttpContextAccessor httpContextAccessor)
  {
    _httpContextAccessor = httpContextAccessor;
  }

  public bool IsUserLoggedIn()
  {
    var context = _httpContextAccessor.HttpContext;
    return context.User.Identities.Any(x => x.IsAuthenticated);
  }
}
```

**Reply by petebrieso** [reply]
I don't see here how you're making the values available to a typescript unit. I don't want to be calling an endpoint to get this value either, just looking for a cleaner way to make my base value available directly inside my typescript code, and does Serenity offer a pattern to achieve it?
I also know I can add values to the user object that originate in appsettings, but it's not right for my case. 
Is there something close to how `LookupScript` adds values to a list that's readable in ts?

**Comment by minhhungit** [comment]
use ContentResult to create a javascript then embed it using url action in your master page (layout.cshtml) 

```cshtml
<script type="text/javascript" src="@Url.Action("InitialData", "ServerSettingPage", new {requestId = Guid.NewGuid()})"></script>
<script type="text/javascript">
  window.MrGlobalSetting = (function () {
      function loadGlobalSettings() {
          if (typeof g_YOURNAME_SETTINGS !== 'undefined') {
              return g_YOURNAME_SETTINGS ;
          }
          return null;
      }
  
      var faker = {
          // options
          settings: null,
  
          // inital method
          init: function () {
              // load settings
              faker.settings = loadGlobalSettings();
          }
      };
  
      return faker;
  }());

  MrGlobalSetting.init();
</script>
```


```csharp
[RoutePrefix("YourName/Settings"), Route("{action=index}")]
public class ServerSettingPageController : Controller
{
    [Authorize]
    public ContentResult InitialData()
    {
        var javascripts = new List<string>
        {
            "var g_YOURNAME_SETTINGS = {"ExpireTime": 12 };" // load setting from appsetting and build to string here
        };

        var result = $"{string.Join("\n\n", javascripts.Select(x => x.TrimEnd(';') + ";"))}";

        return Content(result, "text/javascript");
    }
}
```

then at javascript/typescript code you can use `g_YOURNAME_SETTINGS` variable

```javascript
let expiredTimeHours = MrGlobalSetting.settings.ExpireTime;
```

Note that you also need to define d.ts to make your setting strongly-type

```typescript
declare namespace YOUR_NAMESPACE{
    interface IGlobalSetting {
        settings: YOURNAMESPACE.Common.Settings.ServerSettingModel;
        init(): void;
    }
}

declare var MrGlobalSetting: YOUR_NAMESPACE.IGlobalSetting ;
```

Let me know if it is not clear

**Reply by petebrieso** [reply]
thanks, it's clear what you're doing but it still uses the global space in javascript, and populating it directly from the injected `IConfiguration` object is less code.
I was hoping for a something like `[ScriptInclude]` as a decorator to generated a ts hook. Is that doable?

**Reply by minhhungit** [reply]
I think you can limit what can be show based on logged user
